contracts/integrations/test/aggregation/fill_test.ts

```diff
@@ -0,0 +1,218 @@
+                continue;
+            }
+            if (uniqueOrders.findIndex(o => isSameOrder(order, o)) === -1) {
+                uniqueOrders.push(order);
+            }
+        }
+        return uniqueOrders;
+    }
+
+    function isSameOrder(a: SignedOrder, b: SignedOrder): boolean {
+        for (const [k, v] of Object.entries(a)) {
+            if (k in (b as any)) {
+                if (BigNumber.isBigNumber(v) && !v.eq((b as any)[k])) {
+                    return false;
+                }
+                if (v !== (b as any)[k]) {
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
+
+    function toTokenUnits(symbol: string, weis: Numberish): BigNumber {
+        return new BigNumber(weis).div(new BigNumber(10).pow(tokens[symbol].decimals));
+    }
+
+    function fromTokenUnits(symbol: string, units: Numberish): BigNumber {
+        return new BigNumber(units)
+            .times(new BigNumber(10).pow(tokens[symbol].decimals))
+            .integerValue(BigNumber.ROUND_DOWN);
+    }
+
+    interface MarketOperationResult {
+        makerAssetBalanceBefore: BigNumber;
+        takerAssetBalanceBefore: BigNumber;
+        makerAssetBalanceAfter: BigNumber;
+        takerAssetBalanceAfter: BigNumber;
+        fillResults: FillResults;
+    }
+
+    // Liquidity is low right now so it's possible we didn't have
+    // enough taker assets to cover the orders, so occasionally we'll get incomplete
+    // fills. This function will catch those cases.
+    // TODO(dorothy-zbornak): Remove this special case when liquidity is up.
+    function checkHadEnoughTakerAsset(
+        quote: MarketBuySwapQuote | MarketSellSwapQuote,
+        result: MarketOperationResult,
+    ): boolean {
+        if (result.takerAssetBalanceBefore.gte(quote.worstCaseQuoteInfo.takerAssetAmount)) {
+            return true;
+        }
+        const takerAssetPct = result.takerAssetBalanceBefore
+            .div(quote.worstCaseQuoteInfo.takerAssetAmount)
+            .times(100)
+            .toNumber()
+            .toFixed(1);
+        logUtils.warn(`Could not acquire enough taker asset to complete the fill: ${takerAssetPct}%`);
+        expect(result.fillResults.makerAssetFilledAmount).to.bignumber.lt(quote.worstCaseQuoteInfo.makerAssetAmount);
+        return false;
+    }
+
+    before(async () => {
+        testContract = new TestMainnetAggregatorFillsContract(TEST_CONTRACT_ADDRESS, env.provider, {
+            ...env.txDefaults,
+            gasPrice: GAS_PRICE,
+            gas: 10e6,
+        });
+        swapQuoter = SwapQuoter.getSwapQuoterForStandardRelayerAPIUrl(env.provider, 'https://api.0x.org/sra');
+        // Pool orderbooks because we're desperate for liquidity.
+        orderbooks.push(
+            swapQuoter.orderbook,
+            Orderbook.getOrderbookForPollingProvider({
+                httpEndpoint: 'https://sra.bamboorelay.com/0x/v3',
```

Comment:
Do you know what happens if this endpoint is non-responsive?

Reply:
The tests throw up. This is admittedly a little fragile while we wait for liquidity to come to the party. I'll add some try/catch clauses to handle these cases more gracefully.